### DISCUSSION TITLE
Unable to connect to ClickHouse with SSL

### DISCUSSION THREAD

**Discussion by guillaumemeyer** [plausible/analytics#1391] [Self-Hosted Support] [answered]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Describe the bug

Context:
- Plausible is hosted in an Azure environment
- ClickHouse is hosted on an Azure container, accessible via https://HOST.azurewebsites.net
- Plausible container connectivity to ClickHouse is configured via the environment variable: `ENV CLICKHOUSE_DATABASE_URL=https://HOST.azurewebsites.net/plausible`

When running the Plausible container, logs are indicating a connectivity issue with ClickHouse:
```sh
 docker run -p 8080:8080 org/plausible

Loading plausible..
Starting dependencies..
Starting repos..
create Plausible.Repo database if it doesn't exist
13:32:46.593 [error] Clickhousex.Protocol (#PID<0.179.0>) failed to connect: ** (ErlangError) Erlang error: :closed
13:32:46.593 [error] Clickhousex.Protocol (#PID<0.178.0>) failed to connect: ** (ErlangError) Erlang error: :closed
13:32:47.759 [error] Clickhousex.Protocol (#PID<0.179.0>) failed to connect: ** (ErlangError) Erlang error: :closed
create Plausible.ClickhouseRepo database if it doesn't exist
13:32:47.990 [error] GenServer #PID<0.212.0> terminating
** (UndefinedFunctionError) function :closed.exception/1 is undefined (module :closed is not available)
    :closed.exception([])
    (db_connection 2.2.2) lib/db_connection/connection.ex:87: DBConnection.Connection.connect/2
    (connection 1.0.4) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: nil
13:32:48.004 [error] GenServer #PID<0.214.0> terminating
** (UndefinedFunctionError) function :closed.exception/1 is undefined (module :closed is not available)
    :closed.exception([])
    (db_connection 2.2.2) lib/db_connection/connection.ex:87: DBConnection.Connection.connect/2
    (connection 1.0.4) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: nil
13:32:48.019 [error] GenServer #PID<0.216.0> terminating
** (UndefinedFunctionError) function :closed.exception/1 is undefined (module :closed is not available)
    :closed.exception([])
    (db_connection 2.2.2) lib/db_connection/connection.ex:87: DBConnection.Connection.connect/2
    (connection 1.0.4) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: nil
13:32:48.029 [error] GenServer #PID<0.218.0> terminating
** (UndefinedFunctionError) function :closed.exception/1 is undefined (module :closed is not available)
    :closed.exception([])
    (db_connection 2.2.2) lib/db_connection/connection.ex:87: DBConnection.Connection.connect/2
    (connection 1.0.4) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: nil
** (MatchError) no match of right hand side value: {:error, "killed"}
    (plausible 0.0.1) lib/plausible_release.ex:48: anonymous fn/2 in Plausible.Release.createdb/0
    (elixir 1.11.3) lib/enum.ex:2193: Enum."-reduce/3-lists^foldl/2-0-"/3
    (plausible 0.0.1) lib/plausible_release.ex:47: Plausible.Release.createdb/0
    (stdlib 3.14) erl_eval.erl:680: :erl_eval.do_apply/6
    (elixir 1.11.3) lib/code.ex:341: Code.eval_string_with_error_handling/3
```

While running the same Plausible container using `http` instead of `https` in the `CLICKHOUSE_DATABASE_URL` environment variable works.

### Expected behavior

Plausible should be able to support SSL connectivity to clickhouse

### Environment

```markdown
- OS: N/A
- Browser: N/A
- Browser Version: N/A
```

**Comment by guillaumemeyer** [comment]
Tried to use https in other environments than Azure, and I'm facing the same issue.
Looks like an issue with Clickhousex.Protocol

**Comment by ukutaht** [comment, marked as answer]
SSL must be configured in Clickhouse: https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings/#http-porthttps-port

**Reply by guillaumemeyer** [reply]
Actually our ClickHouse server is exposed through a proxy that already handles SSL. The server responds without errors to https requests and the certificate is valid.
When using `CLICKHOUSE_DATABASE_URL` with http, Plausible is able to access it, not when using https.

**Reply by ukutaht** [reply]
Oh, interesting. I've only ever used http because we're on a private network.

You can try adding `?ssl=true` to the `CLICKHOUSE_DATABASE_URL` parameter. It enables SSL when using Ecto (the db connection library) with Postgres. Might work for Clickhouse as well.

If that doesn't work I can't help you. Maybe you could ask the authors of the clickhouse connection library or Ecto itself.

**Reply by m-col** [reply]
There is an example config with SSL support in the clickhouse ecto repo. Maybe useful? https://github.com/clickhouse-elixir/clickhouse_ecto/blob/bc963d3f53bf4264e6fa2b7345c7e3c3dc49d6ee/examples/config/prod.exs#L22-L44

**Comment by ruslandoga** [comment]
👋 everyone

I tried running clickhouse on digitalocean with caddy in front of it and encountered the same issue.

Seems like `clickhousex` [defaults to `scheme: :http`](https://github.com/plausible/clickhousex/blob/e6b7cc5a1ea7bb26d77af8fdae03952f662469af/lib/clickhousex/protocol.ex#L26), regardless of what's in the url (ecto ignores the scheme). This `:scheme` option is then [used](https://github.com/plausible/clickhousex/blob/e6b7cc5a1ea7bb26d77af8fdae03952f662469af/lib/clickhousex/protocol.ex#L41) for [rebuilding the URL incorrectly](https://github.com/plausible/clickhousex/blob/e6b7cc5a1ea7bb26d77af8fdae03952f662469af/lib/clickhousex/protocol.ex#L217) when sending requests.

**Reply by ruslandoga** [reply]
I see three workarounds:

- a quick one, adding correct `:scheme` to clickhouse repo options https://github.com/plausible/analytics/pull/2427
- modifying plausible's fork to use scheme from the connection string and fallback to options ~~(not possible since ecto doesn't pass original url in options to dbconnection process init)~~ https://github.com/plausible/analytics/pull/2570
- ~~using mint which doesn't need to rebuild the url~~
